lib/kubernetes-deploy/runner_task.rb

```diff
@@ -0,0 +1,132 @@
+      @kubeclient = build_v1_kubeclient(context)
+      @context = context
+    end
+
+    def run(task_template:, command:, args:, env_vars: [])
+      validate_configuration(task_template, @namespace, args)
+
+      raw_template = get_template(task_template, namespace: @namespace)
+      rendered_template = build_pod_template(raw_template, command, args, env_vars)
+
+      KubernetesDeploy.logger.info("Starting task runner pod #{rendered_template.metadata.name}")
+
+      @kubeclient.create_pod(rendered_template)
+    end
+
+    private
+
+    def validate_configuration(task_template, namespace, args)
+      errors = []
+
+      if task_template.blank?
+        errors << "Task template name can't be nil"
+      end
+
+      if args.nil?
+        errors << "Args can't be nil"
+      end
+
+      if namespace.blank?
+        errors << "Namespace can't be empty"
+      end
+
+      raise FatalTaskRunError, "Configuration invalid: #{errors.join(', ')}" unless errors.empty?
+    end
+
+    def get_template(template_name, namespace:)
+      KubernetesDeploy.logger.info("Fetching task runner pod template: '#{template_name}' in namespace: '#{namespace}'")
+
+      pod_template = begin
+        @kubeclient.get_pod_template(template_name, namespace)
+      rescue KubeException => e
+        raise FatalTaskRunError, "No podtemplate found matching name=#{template_name}" \
+          " in namespace=#{namespace}, err: #{e.message}"
+      end
+
+      raw = pod_template.template
+
+    rescue Psych::SyntaxError => e
+      KubernetesDeploy.logger.error(raw.to_hash)
+      raise FatalTaskRunError, "Task template #{template_name} cannot be parsed: #{e.message}"
+    end
+
+    def build_pod_template(base_template, command, args, env_vars)
+      KubernetesDeploy.logger.info("Rendering template for task runner pod")
+
+      rendered_template = base_template.dup
+      rendered_template.kind = 'Pod'
+      rendered_template.apiVersion = 'v1'
+      rendered_template.spec.containers
+        .first.args = args
+
+      container = rendered_template.spec.containers.find { |cont| cont.name == 'task-runner' }
+
+      raise FatalTaskRunError, "Pod spec does not contain a template container called 'task-runner'" if container.nil?
+
+      container.command = command
+      container.env ||= []
+      container.env = container.env.map(&:to_h) + env_vars.map { |env| env.split('=') }.map do |key, value|
+        { name: key, value: value }
+      end
+
+      unique_name = rendered_template.metadata.name + "-" + SecureRandom.hex(8)
+
+      KubernetesDeploy.logger.warn("Name is too long, using '#{unique_name[0..62]}'") if unique_name.length > 63
+      rendered_template.metadata.name = unique_name[0..62]
+      rendered_template.metadata.namespace = @namespace
+
+      validate_pod_spec(rendered_template)
+
+      rendered_template
+    end
+
+    def validate_pod_spec(pod)
+      f = Tempfile.new(pod.metadata.name)
+      f.write recursive_to_h(pod).to_json
```

Comment:
Ugh, this is a pain. I'm thinking it would be much easier/cleaner to use kubectl in place of `get_pod_template` and manipulate the json/hash. It seems like the gem is getting in the way FAR more than it is helping here.

I've encountered similar pains in the feature I'm working on right now (needed to `.to_h.stringify_keys` some things). cc @kirs between this and your simple patch for the 404 error still not being accepted, I'm having second thoughts about favouring the gem over kubectl. 😞